File: app/src/main/java/space/pal/sig/old/view/launches/LaunchesAdapter.java
package space.pal.sig.old.view.launches;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import space.pal.sig.R;
import space.pal.sig.old.model.Launch;
import space.pal.sig.old.repository.dto.LocationDto;
import space.pal.sig.old.repository.dto.RocketDto;
import space.pal.sig.old.repository.dto.SpaceProviderDto;

/**
 * SpaceNow
 * Created by Catalin on 7/19/2020
 **/
public class LaunchesAdapter extends PagedListAdapter<Launch, LaunchesAdapter.ViewHolder> {

    private static final DiffUtil.ItemCallback<Launch> launchItemCallback =
            new DiffUtil.ItemCallback<Launch>() {
                @Override
                public boolean areItemsTheSame(@NonNull Launch oldItem, @NonNull Launch newItem) {
                    return oldItem.getId().equals(newItem.getId());
                }

                @Override
                public boolean areContentsTheSame(@NonNull Launch oldItem, @NonNull Launch newItem) {
                    return oldItem.equals(newItem);
                }
            };
    private static final int TOP = 1;
    private static final int MIDDLE = 2;
    private static final int BOTTOM = 3;
    private static final int TOP_BOTTOM = 4;
    private final SelectLaunchListener listener;

    public LaunchesAdapter(SelectLaunchListener listener) {
        super(launchItemCallback);
        this.listener = listener;
    }

    public Launch getLaunch(int position) {
        return getItem(position);
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0 && position == getItemCount() - 1) {
            return TOP_BOTTOM;
        } else if (position == 0) {
            return TOP;
        } else if (position == getItemCount() - 1) {
            return BOTTOM;
        } else {
            return MIDDLE;
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.launch_view, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.render(getItem(position), getItemViewType(position));
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.launch_image)
        AppCompatImageView image;
        @BindView(R.id.launch_name)
        AppCompatTextView name;
        @BindView(R.id.launch_agency)
        AppCompatTextView agency;
        @BindView(R.id.launch_location)
        AppCompatTextView location;
        @BindView(R.id.launch_date)
        AppCompatTextView date;
        @BindView(R.id.line)
        View line;
        @BindView(R.id.launch_layout)
        ConstraintLayout layout;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        public void render(Launch launch, int viewType) {
            if (launch == null) return;
            RocketDto rocketDto = launch.getRocket();
            if (rocketDto == null) return;
            RecyclerView.LayoutParams params = (RecyclerView.LayoutParams)
                    layout.getLayoutParams();
            switch (viewType) {
                case TOP_BOTTOM:
                    line.setBackgroundResource(R.drawable.launches_line_top_bottom);
                    params.topMargin = 33;
                    break;
                case TOP:
                    line.setBackgroundResource(R.drawable.launches_line_top);
                    params.topMargin = 33;
                    break;
                case BOTTOM:
                    line.setBackgroundResource(R.drawable.launches_line_bottom);
                    params.bottomMargin = 33;
                    break;
                default:
                    line.setBackgroundResource(R.drawable.launches_line_middle);
                    break;
            }
            layout.setLayoutParams(params);
            Picasso.get().load(rocketDto.getImageURL())
                    .centerCrop()
                    .fit()
                    .into(image);
            image.setContentDescription(launch.getName());
            name.setText(launch.getName());
            SpaceProviderDto spaceProvider = launch.getLaunchSpaceProvider();
            agency.setText(spaceProvider != null ? spaceProvider.getName() : "");
            LocationDto locationDto = launch.getLocation();
            if (locationDto != null &&
                    locationDto.getPads() != null &&
                    locationDto.getPads().size() > 0) {
                location.setText(locationDto.getPads().get(0).getName());
            } else {
                location.setText("");
            }
            date.setText(launch.getDate());
        }

        @OnClick(R.id.launch_layout)
        void onClick() {
            listener.onClick(getLaunch(getLayoutPosition()));
        }
    }
}

File: app/src/main/java/space/pal/sig/view/launches/LaunchesAdapter.kt
package space.pal.sig.view.launches

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.squareup.picasso.Picasso
import space.pal.sig.R
import space.pal.sig.databinding.LaunchViewBinding
import space.pal.sig.model.entity.LaunchWithData
import space.pal.sig.util.displayDatetime
import java.util.*
import kotlin.collections.HashMap

/**
 * SpaceNow
 * Created by Catalin on 12/21/2020
 **/
class LaunchesAdapter(
        private val context: Context,
        private val listener: LaunchClickListener
) : RecyclerView.Adapter<LaunchesAdapter.ViewHolder>() {

    companion object {
        private const val TOP = 1
        private const val MIDDLE = 2
        private const val BOTTOM = 3
        private const val TOP_BOTTOM = 4
    }

    private val launchesList = mutableListOf<LaunchWithData>()
    private val imagesMap = HashMap<LaunchWithData, String?>()

    fun submit(launches: List<LaunchWithData>) {
        launchesList.clear()
        launchesList.addAll(launches)
        setupImages(launches)
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(context)
        val binding: LaunchViewBinding = DataBindingUtil
                .inflate(inflater, R.layout.launch_view, parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.render(launchesList[position], getItemViewType(position))
    }

    override fun getItemCount(): Int {
        return launchesList.size
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0 && position == itemCount - 1) {
            TOP_BOTTOM
        } else if (position == 0) {
            TOP
        } else if (position == itemCount - 1) {
            BOTTOM
        } else {
            MIDDLE
        }
    }

    private fun setupImages(launches: List<LaunchWithData>) {
        launches.forEach { launch ->
            val rocket = launch.rocket
            var image: String? = null
            rocket?.let {
                image = when (val imagesSize = rocket.images.size) {
                    0 -> null
                    else -> {
                        val index = Random().nextInt(imagesSize)
                        rocket.images[index]
                    }
                }
            }
            imagesMap[launch] = image
        }
    }

    inner class ViewHolder(
            private val binding: LaunchViewBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun render(launchWithData: LaunchWithData, viewType: Int) {
            binding.launchLayout.setOnClickListener { listener.onClick(launchWithData) }
            val params = binding.launchLayout.layoutParams as RecyclerView.LayoutParams
            when (viewType) {
                TOP_BOTTOM -> {
                    binding.line.setBackgroundResource(R.drawable.launches_line_top_bottom)
                    params.topMargin = 33
                }
                TOP -> {
                    binding.line.setBackgroundResource(R.drawable.launches_line_top)
                    params.topMargin = 33
                }
                BOTTOM -> {
                    binding.line.setBackgroundResource(R.drawable.launches_line_bottom)
                    params.bottomMargin = 33
                }
                else -> binding.line.setBackgroundResource(R.drawable.launches_line_middle)
            }
            binding.launchLayout.layoutParams = params
            val launch = launchWithData.launch
            val rocket = launchWithData.rocket
            val launchpad = launchWithData.launchpad
            val date = Date(launch.dateUnix * 1000).displayDatetime()
            binding.launchName.text = launch.name
            binding.launchDate.text = context.getString(R.string.launch_date_label, date)
            binding.launchAgency.text = context.getString(R.string.space_x_label)
            launchpad?.let {
                binding.launchLocation.text = it.fullName
            }
            rocket?.let {
                binding.launchRocket.text = it.name
            }
            val image: String? = imagesMap[launchWithData]
            image?.let {
                Picasso.get()
                        .load(it)
                        .fit()
                        .centerCrop()
                        .into(binding.launchImage)
            }
        }
    }
}